hooks/checkout

```diff
@@ -2,26 +2,53 @@
 
 set -eu
 
-main() {
-  if [[ -z "${BUILDKITE_REPO:-}" ]]; then
-    echo "error: BUILDKITE_REPO not set"
-    exit 1
-  fi
-  if [[ -z "${BUILDKITE_BRANCH:-}" ]]; then
-    echo "error: BUILDKITE_BRANCH not set"
-    exit 1
-  fi
-  if [[ -z "${BUILDKITE_COMMIT:-}" ]]; then
-    echo "error: BUILDKITE_COMMIT not set"
+# Checks if an env var is set
+# Arguments:
+# $1: var name
+check_set() {
+  local name="$1"
+  if [[ -z "${!name:-}" ]]; then
+    echo "ERROR: ${name} not set"
     exit 1
   fi
+}
 
-  git remote set-url origin "${BUILDKITE_REPO}"
+checkout() {
   git clean -ffxdq
   git fetch -v origin HEAD
   git config remote.origin.fetch
   git fetch -v --prune origin "${BUILDKITE_BRANCH}"
   git checkout -f "${BUILDKITE_COMMIT}"
 }
 
+clean_checkout_dir() {
+  rm -rf "${BUILDKITE_BUILD_CHECKOUT_PATH}"
+  mkdir -p "${BUILDKITE_BUILD_CHECKOUT_PATH}"
+  cd "${BUILDKITE_BUILD_CHECKOUT_PATH}"
+}
+
+clone() {
+  # git clone needs an empty directory
+  clean_checkout_dir
+  git clone "${BUILDKITE_REPO}" .
+}
+
+main() {
+  check_set BUILDKITE_REPO
+  check_set BUILDKITE_BRANCH
+  check_set BUILDKITE_COMMIT
+  check_set BUILDKITE_BUILD_CHECKOUT_PATH
+
+  local checkout_status=0
+  checkout || checkout_status=$?
+
+  if [[ "${checkout_status}" -ne 0 ]]; then
+    echo "checkout failed"
+    # if checkout fails, then clone and try checkout again
```

Comment:
I think this is too aggressive, and part of the reason why the buildkite jobs get stuck in a loop constantly cloning and killing the rate limit with the default checkout logic. Let's check for specific reasons. At the moment we know this will fail if the checkout directory isn't a git repo for example, so let's cover just that case first.

Reply:
yep I was thinking something like that, but wasn't sure if that would cover the cases we have seen from agents before.
Will update.